ext/openssl/openssl_pwhash.c

```diff
@@ -0,0 +1,396 @@
+		zend_long smemlimit = zval_get_long(opt);
+
+		if ((smemlimit < 0) || (smemlimit < MEMLIMIT_MIN) || (smemlimit > (MEMLIMIT_MAX))) {
+			zend_value_error("Memory cost is outside of allowed memory range");
+			return FAILURE;
+		}
+		*memlimit = smemlimit;
+	}
+	if ((opt = zend_hash_str_find(options, "time_cost", strlen("time_cost")))) {
+		zend_long sopslimit = zval_get_long(opt);
+		if ((sopslimit < OPSLIMIT_MIN) || (sopslimit > OPSLIMIT_MAX)) {
+			zend_value_error("Time cost is outside of allowed time range");
+			return FAILURE;
+		}
+		*opslimit = sopslimit;
+	}
+	if ((opt = zend_hash_str_find(options, "threads", strlen("threads"))) && (zval_get_long(opt) != 1)) {
+		zend_long sthreads = zval_get_long(opt);
+		if ((sthreads < THREADS_MIN) || (sthreads > THREADS_MAX)) {
+			zend_value_error("Invalid number of threads");
+			return FAILURE;
+		}
+		*threads = sthreads;
+	}
+	return SUCCESS;
+}
+
+static bool php_openssl_argon2_compute_hash(
+	const char *algo,
+	uint32_t version, uint32_t memlimit, uint32_t opslimit, uint32_t threads,
+	const char *pass, size_t pass_len,
+	const unsigned char *salt, size_t salt_len,
+	unsigned char *hash, size_t hash_len)
+{
+	OSSL_PARAM params[7], *p = params;
+	EVP_KDF *kdf = NULL;
+	EVP_KDF_CTX *kctx = NULL;
+	bool ret = false;
+
+
+	if (threads > 1) {
+		if (OSSL_set_max_threads(NULL, threads) != 1) {
+			goto fail;
+		}
+	}
+	p = params;
+	*p++ = OSSL_PARAM_construct_uint32(OSSL_KDF_PARAM_THREADS,
+		&threads);
+	*p++ = OSSL_PARAM_construct_uint32(OSSL_KDF_PARAM_ARGON2_LANES,
+		&threads);
+	*p++= OSSL_PARAM_construct_uint32(OSSL_KDF_PARAM_ITER,
+		&opslimit);
+	*p++ = OSSL_PARAM_construct_uint32(OSSL_KDF_PARAM_ARGON2_MEMCOST,
+		&memlimit);
+	*p++ = OSSL_PARAM_construct_octet_string(OSSL_KDF_PARAM_SALT,
+		(void *)salt, salt_len);
+	*p++ = OSSL_PARAM_construct_octet_string(OSSL_KDF_PARAM_PASSWORD,
+		(void *)pass, pass_len);
+    *p++ = OSSL_PARAM_construct_end();
+
+    if ((kdf = EVP_KDF_fetch(NULL, algo, NULL)) == NULL) {
+		goto fail;
+	}
+	if ((kctx = EVP_KDF_CTX_new(kdf)) == NULL) {
+		goto fail;
+	}
+    if (EVP_KDF_derive(kctx, hash, hash_len, params) != 1) {
+		zend_value_error("Unexpected failure hashing password");
+		goto fail;
+	}
+
+	ret = true;
+
+fail:
+    EVP_KDF_free(kdf);
+    EVP_KDF_CTX_free(kctx);
+
+	if (threads > 1) {
+	    OSSL_set_max_threads(NULL, 0);
+	}
+	return ret;
+}
+
+static zend_string *php_openssl_argon2_hash(const zend_string *password, zend_array *options, const char *algo)
+{
+	uint32_t opslimit, memlimit, threads, version = ARGON_VERSION;
+	zend_string *digest = NULL, *salt64 = NULL, *hash64 = NULL;
+	unsigned char hash[HASH_SIZE+1], salt[SALT_SIZE+1];
+
+	if ((ZSTR_LEN(password) >= UINT32_MAX)) {
+		zend_value_error("Password is too long");
+		return NULL;
+	}
+	if (get_options(options, &memlimit, &opslimit, &threads) == FAILURE) {
+		return NULL;
+	}
+	if (FAILURE == php_random_bytes_throw(salt, SALT_SIZE)) {
```

Comment:
I think we should use OpenSSL RAND api instead so this is also ok for fips in the future.